File: soya/src/pro_181188.ts
function solution(targets) {
  let answer = 0;
  targets.sort((a, b) => a[0] - b[0]); //오름차순 정렬
  let curS = targets[0][0];
  let curE = targets[0][1]; //작은걸로 유지

  for (let i of targets) {
    let s = i[0];
    let e = i[1];
    //s 가 변화가 없는 경우
    if (curS === s && curE <= e) continue;
    if (curS === s && curE > e) {
      curE = e;
      continue;
    }
    //s 가 변화하는경우
    curS = s;
    if (curE <= curS) {
      answer++; // 앞에 애들 요격시키기.
      curE = e;
      continue;
    }
    if (curE > e) {
      curE = e;
    }
  }
  answer++;
  return answer;
}
let targets = [
  [4, 5],
  [4, 8],
  [10, 14],
  [11, 13],
  [5, 12],
  [3, 7],
  [1, 4],
];

console.log(solution(targets));

File: java/JaeHyeon/pro_181188.java
import java.util.*;

class Solution {
  public int solution(int[][] targets) {
    Arrays.sort(targets, (o1, o2) -> {
      if (o1[0] == o2[0]) {
        return o1[1] - o2[1];
      }
      return o1[0] - o2[0];
    });

    // for (int i = 0; i < targets.length; i++) {
    //     System.out.print(targets[i][0] + ", " + targets[i][1] + "\n");
    // }

    int check = targets[0][1];

    int count = 0;

    for (int i = 1; i < targets.length; i++) {

      if (targets[i][0] >= check) {
        count++;
        check = targets[i][1];
        // System.out.println(check + " " + count);
      } else {

        // 기준으로 잡은 선에 완전히 겹치면
        // check값을 갱신해주지만,
        // 그렇지 않으면, 갱신해주면 안됨
        if (check >= targets[i][1]) {
          check = targets[i][1];
        }
      }
    }

    return count+1;
  }
}
